rtmr/extend/extend.go

```diff
@@ -0,0 +1,199 @@
+)
+
+// RtmrExtend is a struct that represents a rtmr entry in the configfs.
+type RtmrExtend struct {
+	RtmrIndex int
+	entry     *configfsi.TsmPath
+	client    configfsi.Client
+}
+
+// Remove the null bytes from the c-string
+func convertCstring(b []byte) string {
+	return string(bytes.Trim(b, "\x00"))
+}
+
+func (r *RtmrExtend) attribute(subtree string) string {
+	a := *r.entry
+	a.Attribute = subtree
+	return a.String()
+}
+
+// ExtendDigest extends the measurement to the rtmr with the given hash.
+func (r *RtmrExtend) ExtendDigest(hash []byte) error {
+	if (r.RtmrIndex != 2) && (r.RtmrIndex != 3) {
+		return fmt.Errorf("invalid rtmr index %d, userspace can only extend to rtmr2 or rtmr3", r.RtmrIndex)
+	}
+	if err := r.client.WriteFile(r.attribute(TsmRtmrDigest), hash); err != nil {
+		return fmt.Errorf("could not write report %s: %v", TsmPathIndex, err)
+	}
+	return nil
+}
+
+// SetRtmrIndex sets a configfs rtmr entry to the given index.
+// It reports an error if the index cannot be written or the rtmr_tcg map does not match.
+func (r *RtmrExtend) SetRtmrIndex() error {
+	indexBytes := []byte(strconv.Itoa(r.RtmrIndex)) // Convert index to []byte
+	indexPath := r.attribute(TsmPathIndex)
+	if err := r.client.WriteFile(indexPath, indexBytes); err != nil {
+		return fmt.Errorf("could not write index %s: %v", indexPath, err)
+	}
+	if err := r.ValidateRtmrIndex(); err != nil {
+		return fmt.Errorf("the tcg_map is invalid %s: %v", indexPath, err)
+	}
+	return nil
+}
+
+// ValidateRtmrIndex checks if the rtmr to PCR map is valid.
+// It returns an error if the rtmr_tcg map does not match the expected value.
+func (r *RtmrExtend) ValidateRtmrIndex() error {
+	if (r.client == nil) || (r.entry == nil) {
+		return fmt.Errorf("RtmrExtend is not initialized")
+	}
+	data, err := r.client.ReadFile(r.attribute(TsmPathTcgMap))
+	if err != nil {
+		return fmt.Errorf("could not read  %s: %v", TsmPathTcgMap, err)
+	}
+	var rtmrPcrMaps = map[int]string{
+		0: "1,7\n",
+		1: "2-6\n",
+		2: "8-15\n",
+		3: "\n",
+	}
+
+	inputMap := convertCstring(data)
+	if inputMap != rtmrPcrMaps[r.RtmrIndex] {
+		return fmt.Errorf("tcg map error. expect:%s get:%s", rtmrPcrMaps[r.RtmrIndex], string(data))
+	}
+	return nil
+}
+
+// SearchRtmrInterface searches for an rtmr entry in the configfs.
+func SearchRtmrInterface(client configfsi.Client, index int) (*RtmrExtend, error) {
+	root := TsmRtmrPrefix
+	out := RtmrExtend{}
+	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() {
+			p, err := configfsi.ParseTsmPath(path)
+			if err != nil {
+				return err
+			}
+			r := RtmrExtend{
+				RtmrIndex: index,
+				entry:     &configfsi.TsmPath{Subsystem: RtmrSubsystem, Entry: p.Entry},
+				client:    client,
+			}
+			if r.ValidateRtmrIndex() == nil {
+				out = r
+				return nil
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	if out.ValidateRtmrIndex() != nil {
```

Comment:
You use err in the block but it's always nil. Do you mean
```suggestion
	if err := out.ValidateRtmrIndex(); err != nil {
```